File: agent-service/backend/middleware/verify_signature.py
import hmac
import hashlib
from fastapi import Request, HTTPException
from starlette.middleware.base import BaseHTTPMiddleware
import json
import os

from backend.config import main as config

# Load your secrets (simulate config.HMAC_*_SECRET)
HMAC_SECRETS = {
    "agent": config.HMAC_AGENT_KEY,
    "user": config.HMAC_USER_KEY,
}

async def verify_signature(request: Request):
    headers = request.headers

    signature = headers.get("pente-signature")
    origin = headers.get("pente-origin")
    validate = headers.get("pente-validate")
    timestamp = headers.get("pente-timestamp")
    

    if not all([signature, origin, validate, timestamp]):
        raise HTTPException(status_code=403, detail="Signature Invalid")

    body_bytes = await request.body()
    try:
        body_json = json.loads(body_bytes.decode("utf-8"))
    except json.JSONDecodeError:
        body_json = {}

    query_params = dict(request.query_params)

    if validate == "query":
        validator = json.dumps(query_params, separators=(',', ':')) + timestamp
    elif validate == "both":
        validator = json.dumps(query_params, separators=(',', ':')) + json.dumps(body_json, separators=(',', ':')) + timestamp
    else:
        validator = json.dumps(body_json, separators=(',', ':')) + timestamp

    key = HMAC_SECRETS.get(origin)
    if not key:
        raise HTTPException(status_code=403, detail="Signature Invalid")

    generated_signature = hmac.new(
        key.encode("utf-8"),
        validator.encode("utf-8"),
        hashlib.sha256
    ).hexdigest()

    if not hmac.compare_digest(generated_signature, signature):
        raise HTTPException(status_code=403, detail="Signature Invalid")

File: user-service/middlewares/verifySignature.js
import crypto from 'crypto'
import config from '@/config'
import isEmpty from 'is-empty'

/**
 * Important Params,
 *
 * @field PENTE-SIGNATURE - header - signature
 * @field PENTE-ORIGIN - request origin - type of service [ agent | user ]
 * @field PENTE-VALIDATE - validator type - used for signature verification [ body | query | both <query + body> ]
 * @field PENTE-TIMESTAMP - timestamp of the placed request - Its required for signature authenticity
 *
 */
export const verifySignature = async (req, res, next) => {
    try {
        const { headers, body, query } = req
        
        const signature = headers['pente-signature']
        const origin = headers['pente-origin']
        const validate = headers['pente-validate']
        const timestamp = headers['pente-timestamp']

        if (isEmpty(signature) || isEmpty(origin) || isEmpty(validate) || isEmpty(timestamp)) {
            return res.status(403).json({ success: false, message: 'Signature Invalid' })
        }

        let validator = JSON.stringify(body) + timestamp

        if (validate === 'query') {
            validator = JSON.stringify(query) + timestamp
        } else if (validate === 'both') {
            validator = JSON.stringify(query) + JSON.stringify(body) + timestamp
        }

        let key = ''

        if (origin === 'agent') {
            key = config.HMAC_AGENT_KEY
        } else if (origin === 'user') {
            key = config.HMAC_USER_KEY
        } else {
            return res.status(403).json({ success: false, message: 'Invalid Origin' })
        }

        const generatedSignature = crypto.createHmac('SHA256', Buffer.from(key, 'utf-8'))
        generatedSignature.update(validator)

        const latestSignature = generatedSignature.digest('hex')
        const isValid = crypto.timingSafeEqual(Buffer.from(latestSignature, 'hex'), Buffer.from(signature, 'hex'))

        if (!isValid) {
            return res.status(403).json({ success: false, message: 'Signature Invalid' })
        } else {
            return next()
        }
    } catch (error) {
        console.log(error)
        return res.status(500).json({ success: false, message: 'Something went wrong' })
    }
}
